include/boost/url/detail/any_allocator.hpp

```diff
@@ -0,0 +1,251 @@
+        is_equal(const any_allocator_base::base&
+                p) = 0;
+    };
+
+    template <typename Allocator>
+    struct impl
+        : public base
+        , private empty_value<
+              boost::allocator_rebind_t<Allocator,
+                  char>,
+              0>
+    {
+        using allocator_type = boost::
+            allocator_rebind_t<Allocator, char>;
+
+        explicit impl(allocator_type const& a)
+            : empty_value<allocator_type, 0>(
+                  empty_init_t(), a)
+        {
+            type_id = get_type_id<allocator_type>();
+        }
+
+        void*
+        allocate(std::size_t n,
+            std::size_t size,
+            std::size_t align) override
+        {
+            if(align == 1)
+                return allocator().allocate(n);
+            else
+                return allocate_aligned(
+                    n, size, align);
+        }
+
+        void
+        deallocate(void* p,
+            std::size_t n,
+            std::size_t size,
+            std::size_t align) override
+        {
+            if(align == alignof(char))
+                allocator().deallocate(
+                    reinterpret_cast<char*>(p),
+                    n);
+            else
+                deallocate_aligned(
+                    p, n, size, align);
+        }
+
+        bool
+        is_equal(
+            const any_allocator_base::base& other)
+            override
+        {
+            const impl* other_impl = static_cast<
+                const impl*>(&other);
+            return allocator() ==
+                other_impl->allocator();
+        };
+
+        const allocator_type&
+        allocator() const
+        {
+            return empty_value<allocator_type,
+                0>::get();
+        }
+
+        allocator_type&
+        allocator()
+        {
+            return empty_value<allocator_type,
+                0>::get();
+        }
+
+#if BOOST_WORKAROUND(BOOST_GCC, < 40900)
+        using max_align_t = ::max_align_t;
+#else
+        using max_align_t = std::max_align_t;
+#endif
+
+        static std::size_t
+        ceil_div(std::size_t n, std::size_t d)
+        {
+            return (n + d - 1) / d;
+        }
+
+        template <std::size_t N>
+        typename std::enable_if<
+            alignof(max_align_t) <= N,
+            void*>::type
+        try_allocate_aligned(std::size_t n,
+            std::size_t size,
+            std::size_t align)
+        {
+            using aligned_t = max_align_t;
+            using aligned_allocator_t = boost::
+                allocator_rebind_t<Allocator,
+                    aligned_t>;
+            return aligned_allocator_t(
+                allocator())
+                .allocate(
+                    n * ceil_div(size, align));
+        }
+
+        template <std::size_t N>
+            typename std::enable_if <
+            N< alignof(max_align_t), void*>::type
+            try_allocate_aligned(std::size_t n,
```

Comment:
Something weird is going on here with the margin

Reply:
Should it be larger? That was my fault. It was trying to do that triangle and trapeziums, etc...